### PR TITLE
Misc infrastructure fixes

### PR DESCRIPTION
**Pull request by jaredpar**
Porting a couple of infrastructure fixes to issues I found while working on our VSTS support:

1. Improve the robustness of our determinism tests
1. Improve the robustness of our Mutex creation in our VBCSCompiler tests

### CODE REVIEW

**Comment by jaredpar** [comment]
CC @dotnet/roslyn-infrastructure for review

**Comment by sharwell** [comment]
💭 It sounded like the determinism tests might benefit from a clean before **subst**. The file write lists use absolute paths, so the relocation of folders would leave the clean as incomplete and (IMO) not representative of build goals.

File: build/scripts/test-determinism.ps1 [hunk 1]
```diff
@@ -12,26 +13,25 @@ $ErrorActionPreference = "Stop"
...
-function Run-Build([string]$rootDir, [switch]$restore = $false, [string]$logFile = $null) {
+function Run-Build([string]$rootDir, [string]$logFile = $null) {
     Push-Location $rootDir
     try {
 
         # Clean out the previous run
         Write-Host "Cleaning the Binaries"
-        Exec-Console $msbuild "/nologo /v:m /nodeReuse:false /t:clean Roslyn.sln" 
+        Remove-Item -Recurse (Get-ConfigDir $rootDir) 
+        Remove-Item -Recurse (Get-ObjDir $rootDir)
```
Comment (jaredpar) [inline build/scripts/test-determinism.ps1]: @sharwell this is the clean you're looking for. Before every build the entire Obj and bin directories are nuked. Previously I was using `/t:clean` and that's woefully insufficient. It barely deletes anything.

File: build/scripts/test-determinism.ps1 [hunk 2]
```diff
@@ -1,6 +1,7 @@
 [CmdletBinding(PositionalBinding=$false)]
 param ( [string]$bootstrapDir = "",
-        [switch]$debugDeterminism = $false)
+        [switch]$release = $false,
+        [string]$altRootDrive = "q:")
```
Comment (jasonmalinowski) [inline build/scripts/test-determinism.ps1]: Please document what "altRootDrive" is.

Comment (jasonmalinowski) [reply build/scripts/test-determinism.ps1]: (there's a good comment at the use site 200 lines away...might be good to put it here at the top too.)

**Pull request merged**